### PR TITLE
Package and CP2KMM upgrades

### PR DESCRIPTION
**Pull request by BvB93**
See https://github.com/SCM-NV/qmflows/issues/159.

Major
-----
* Updated the ``Warning`` management in QMFlows. All issued warnings are now instances of ``QMFlows_Warning`` or one of its subclasses.
* Updated the general structure of ``cp2k_warnings``.
* Added warnings for ``CP2KMM``.

Minor
------
* Added a dedicated module for type hints.
* Changed a number of absolute import paths to relative ones.
* Added Sphinx-readable comments to the QMFlows templates.

### CODE REVIEW

**Comment by BvB93** [comment]
@felipeZ I expanded the warning parsing and management in ``warnings_qmflows.py``, as it was a bit too limited for what I was trying to implement. Instead of a [``Warning``](https://docs.python.org/3/library/exceptions.html#Warning) type, the values of ``cp2k_warnings`` now consist of a [``namedtuple``](https://docs.python.org/3/library/typing.html#typing.NamedTuple) with three fields: 

* ``warn_type``: The relevant [``Warning``](https://docs.python.org/3/library/exceptions.html#Warning) type.
* ``parser``: The [``pyparsing.ParserElement``](https://pyparsing-docs.readthedocs.io/en/latest/pyparsing.html#pyparsing.ParserElement) instance for finding the warning message.
* ``func``: A callable for post-processing the warning message. A warning will be skipped if these functions return ``None``.

Any thoughts on the proposed changes thus far?

https://github.com/SCM-NV/qmflows/blob/5567b945b8c69cfcb2b7ad97af67637bc9be49fb/src/qmflows/warnings_qmflows.py#L67-L85

File: src/qmflows/type_hints.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
+"""A module with type-hint related objects used throughout QMFlows.
```
Comment (felipeZ) [inline src/qmflows/type_hints.py]: This module makes things a lot more clear

File: src/qmflows/warnings_qmflows.py [hunk 2]
```diff
@@ -1,14 +1,85 @@
...
+    ret = '\n'.join(i for i in msg_gen if all(j not in i for j in skip))
+    return ret or None
+
+
+def _return_msg(msg: str) -> str:
+    """Return the passed *msg*."""
+    return msg
+
+
+cp2k_warnings: Mapping[str, ParseWarning] = {
```
Comment (felipeZ) [inline src/qmflows/warnings_qmflows.py]: I think that this functionality is flexible and robust to handle the warning that we have encountered so far. Also, it allows run for further extension, so I think that it is a very good idea in general

File: src/qmflows/warnings_qmflows.py [hunk 3]
```diff
@@ -1,14 +1,85 @@
...
+class Parameter_Warning(QMFlows_Warning):
+    """Warning class for classical forcefield parameters."""
+
+
+class Charge_Warning(Parameter_Warning):
+    """Warning class for charges in classical forcefields."""
+
+
+def _eval_charge(msg: str, tolerance: float = 0.1) -> Optional[str]:
+    """Check of the total moleculair charge is integer within a given *tolerance*."""
```
Comment (felipeZ) [inline src/qmflows/warnings_qmflows.py]: small typo on `moleculair` :)

File: src/qmflows/templates/templates.py [hunk 4]
```diff
@@ -261,7 +263,7 @@
             basis: sto_sz
 """,    Loader=yaml.FullLoader))
 
-# Frequencies template
+#: Templates for frequency analyses calculations.
```
Comment (felipeZ) [inline src/qmflows/templates/templates.py]: Good idea to add documentation for these variables

File: src/qmflows/parsers/cp2KParser.py [hunk 5]
```diff
@@ -39,36 +41,40 @@
...
+        for msg_ret, warn_type in iterator:
+            key = named_tup.func(msg_ret)
+            if key is not None:
+                warnings[key] = warn_type
 
-    if not warnings:
-        return None
-    else:
-        return warnings
+    return warnings or None
```
Comment (felipeZ) [inline src/qmflows/parsers/cp2KParser.py]: Beter :+1:

File: src/qmflows/packages/packages.py [hunk 6]
```diff
@@ -240,7 +239,7 @@ def _unpack_results(self) -> None:
         # Failed to find or unpickle the .dill file; issue a warning
         if file_exc is not None:
             self._results = None
-            warn(f"{file_exc}, setting value to 'None'")
+            warn(f"{file_exc}, setting value to 'None'", category=QMFlows_Warning)
```
Comment (felipeZ) [inline src/qmflows/packages/packages.py]: I like that having a `QMFlows_Warning` allows handling the warning on the client code (e.g. qmflows-namd) if we want to

File: src/qmflows/packages/cp2k_package.py [hunk 7]
```diff
@@ -8,11 +8,13 @@
...
-from ..warnings_qmflows import cp2k_warnings
-from .packages import (Package, Result, package_properties,
-                       parse_output_warnings, WarnMap)
+from ..warnings_qmflows import cp2k_warnings, Key_Warning
+from ..type_hints import WarnMap
+
+__all__ = ['cp2k']
 
 # ====================================<>=======================================
 charge_dict: Dict[str, int] = {
```
Comment (felipeZ) [inline src/qmflows/packages/cp2k_package.py]: I am going to replace this dictionary with [mendeleev](https://pypi.org/project/mendeleev/)

**Comment by BvB93** [comment]
Things are ready to get pushed.

File: src/qmflows/packages/cp2k_package.py (hunk 7 above)
Comment (BvB93) [reply src/qmflows/packages/cp2k_package.py]: Shall I remove it in this branch?

File: src/qmflows/warnings_qmflows.py (hunk 3 above)
Comment (BvB93) [reply src/qmflows/warnings_qmflows.py]: Good catch.

File: src/qmflows/packages/cp2k_package.py (hunk 7 above)
Comment (felipeZ) [reply src/qmflows/packages/cp2k_package.py]: I did it on my branch :)

File: src/qmflows/common.py [hunk 8]
```diff
@@ -16,3 +22,9 @@
 InfoMO = namedtuple("InfoMO", ("eigenVals", "coeffs"))
 InputKey = namedtuple("InpuKey", ("name", "args"))
 MO = namedtuple("MO", ("coordinates", "cgfs", "coefficients"))
+
+
+class ParseWarning(NamedTuple):
```
Comment (felipeZ) [inline src/qmflows/common.py]: Great :+1:

**Pull request merged**